src/main.cpp

```diff
@@ -2772,44 +2773,111 @@ void static UpdateTip(CBlockIndex *pindexNew) {
     if (!IsInitialBlockDownload())
     {
         int nUpgraded = 0;
+        bool upgradedEval = false;  // bip135 added
         const CBlockIndex* pindex = chainActive.Tip();
-        for (int bit = 0; bit < VERSIONBITS_NUM_BITS; bit++) {
-            WarningBitsConditionChecker checker(bit);
-            ThresholdState state = checker.GetStateFor(pindex, chainParams.GetConsensus(), warningcache[bit]);
-            if (state == THRESHOLD_ACTIVE || state == THRESHOLD_LOCKED_IN) {
-                if (state == THRESHOLD_ACTIVE) {
-                    strMiscWarning = strprintf(_("Warning: unknown new rules activated (versionbit %i)"), bit);
-                    if (!fWarned) {
-                        AlertNotify(strMiscWarning);
-                        fWarned = true;
+        // bip135 begin
```

Comment:
WarningBitsConditionerChecker was used to check for activation of unknown bits.

With BIP135, it is no longer possible to do that - the best we can do is report on more general events such as those tracked in UnknownForkData . So I had to generalize it, and it seemed useful to add in some thresholds such as 50%, 70%, 95% although they are pretty arbitrary, and the implementation could be reduced somewhat by removing some of them.

I agree that this new checker code should be moved into its own function - will look at improving that.